apps/taxes/src/pages/api/webhooks/order-created.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  OrderCreatedEventSubscriptionFragment,
+  OrderStatus,
+  UntypedOrderCreatedSubscriptionDocument,
+  UpdateMetadataDocument,
+  UpdateMetadataMutation,
+  UpdateMetadataMutationVariables,
+} from "../../../../generated/graphql";
+import { saleorApp } from "../../../../saleor-app";
+import { createLogger } from "../../../lib/logger";
+import { getActiveTaxProvider } from "../../../modules/taxes/active-tax-provider";
+import { createClient } from "../../../lib/graphql";
+import { Client } from "urql";
+import { WebhookResponseFactory } from "../../../modules/app/webhook-response-factory";
+
+export const config = {
+  api: {
+    bodyParser: false,
+  },
+};
+
+type OrderCreatedPayload = Extract<
+  OrderCreatedEventSubscriptionFragment,
+  { __typename: "OrderCreated" }
+>;
+
+export const orderCreatedAsyncWebhook = new SaleorAsyncWebhook<OrderCreatedPayload>({
+  name: "OrderCreated",
+  apl: saleorApp.apl,
+  event: "ORDER_CREATED",
+  query: UntypedOrderCreatedSubscriptionDocument,
+  webhookPath: "/api/webhooks/order-created",
+});
+
+// * This is the key that we use to store the provider order id in the Saleor order metadata.
+export const PROVIDER_ORDER_ID_KEY = "externalId";
+
+// * We need to store the provider order id in the Saleor order metadata so that we can
+// * update the provider order when the Saleor order is fulfilled.
+async function updateOrderMetadataWithExternalId(
+  client: Client,
+  orderId: string,
+  externalId: string
+) {
+  const { error } = await client
+    .mutation<UpdateMetadataMutation>(UpdateMetadataDocument, {
+      id: orderId,
+      input: [
+        {
+          key: PROVIDER_ORDER_ID_KEY,
+          value: externalId,
+        },
+      ],
+    } as UpdateMetadataMutationVariables)
+    .toPromise();
+
+  if (error) {
+    throw error;
+  }
+
+  return { ok: true };
+}
+
+export default orderCreatedAsyncWebhook.createHandler(async (req, res, ctx) => {
+  const logger = createLogger({ event: ctx.event });
+  const { payload, authData } = ctx;
+  const { saleorApiUrl, token } = authData;
+  const webhookResponse = new WebhookResponseFactory(res);
+
+  logger.info({ payload }, "Handler called with payload");
+
+  try {
+    const appMetadata = payload.recipient?.privateMetadata ?? [];
+    const channelSlug = payload.order?.channel.slug;
+    const activeTaxProvider = getActiveTaxProvider(channelSlug, appMetadata);
+
+    if (!activeTaxProvider.ok) {
+      logger.error({ error: activeTaxProvider.error }, "getActiveTaxProvider error");
+      logger.info("Returning no data");
+      return webhookResponse.failureNoRetry(activeTaxProvider.error);
+    }
+
+    logger.info({ activeTaxProvider }, "Fetched activeTaxProvider");
+    const taxProvider = activeTaxProvider.data;
+
+    // todo: figure out what fields are needed and add validation
+    if (!payload.order) {
+      logger.error("Insufficient order data");
+      return res.status(200).json({ success: false });
```

Comment:
I assume you want to replace it with your "error factory" everywhere

Reply:
done